source/s3_auto_ranged_get.c

```diff
@@ -325,97 +401,195 @@ static int s_s3_auto_ranged_get_prepare_request(
+                    &total_object_size)) {
 
-    /* Check if this was the first part and if it was successful. For a ranged-get request, the first part number will
-     * be 1. For an empty file request, the part number will be 0.*/
-    if (error_code == AWS_ERROR_SUCCESS && request->part_number <= 1) {
-        uint64_t total_object_size = 0;
-        if (request->request_tag == AWS_S3_AUTO_RANGE_GET_REQUEST_TYPE_PART) {
-            struct aws_byte_cursor content_range_header_value;
+                AWS_LOGF_ERROR(
+                    AWS_LS_S3_META_REQUEST,
+                    "id=%p Could not find content-range header for request %p",
+                    (void *)meta_request,
+                    (void *)request);
+                break;
+            }
+
+            result = AWS_OP_SUCCESS;
+            break;
+        case AWS_S3_AUTO_RANGE_GET_REQUEST_TYPE_PART:
+            AWS_ASSERT(request->part_number == 1);
+
+            if (error_code != AWS_ERROR_SUCCESS) {
+                /* If we hit an empty file while trying to discover the object-size via part, then this request failure
+                 * is as designed. */
+                if (s_check_empty_file_download_error(request)) {
+                    AWS_LOGF_DEBUG(
+                        AWS_LS_S3_META_REQUEST,
+                        "id=%p Detected empty file with request %p. Sending new request without range header.",
+                        (void *)meta_request,
+                        (void *)request);
+
+                    total_object_size = 0ULL;
+                    total_content_length = 0ULL;
+
+                    result = AWS_OP_SUCCESS;
+                } else {
+                    /* Otherwise, resurface the error code. */
+                    aws_raise_error(error_code);
+                }
+                break;
+            }
+
+            AWS_ASSERT(request->send_data.response_headers != NULL);
+
+            /* Parse the object size from the part response. */
+            if (aws_s3_parse_content_range_response_header(
+                    meta_request->allocator, request->send_data.response_headers, NULL, NULL, &total_object_size)) {
 
-            if (aws_http_headers_get(
-                    request->send_data.response_headers, g_content_range_header_name, &content_range_header_value)) {
                 AWS_LOGF_ERROR(
                     AWS_LS_S3_META_REQUEST,
-                    "id=%p Could not find content range header for request %p",
+                    "id=%p Could not find content-range header for request %p",
                     (void *)meta_request,
                     (void *)request);
 
-                error_code = AWS_ERROR_S3_MISSING_CONTENT_RANGE_HEADER;
-                goto error_encountered;
+                break;
             }
 
-            uint64_t range_start = 0;
-            uint64_t range_end = 0;
+            /* When discovering the object size via first-part, the object range is the entire object. */
```

Comment:
No, the content-range header is what gives us the whole object size.  In this path, we should always have a content-range header for a successful request because we specified a range header.

Reply:
So, from my understanding, we will use `head` request when the `Content-length` is set by initial message. Otherwise, the object range will be entire object. And I think it's better to have this behavior documented. `~assuming no Content-Range header set~ -> assuming no Content-Range header set by initial message`

Reply:
We document it around the spot that we initiate the behavior: https://github.com/awslabs/aws-c-s3/blob/range_get_integration/source/s3_auto_ranged_get.c#L184

We don't branch on the header at this particular spot, so I think it would be awkward to put that documentation here too.